modules/fedimint-ln-client/src/receive.rs

```diff
@@ -0,0 +1,385 @@
+        let invoice = self.confirmed_invoice.invoice.clone();
+        let timeout = invoice.expiry_time();
+        let keypair = self.confirmed_invoice.keypair;
+        let global_context = global_context.clone();
+        vec![
+            StateTransition::new(
+                Self::await_incoming_contract_account(invoice, global_context.clone()),
+                move |dbtx, contract, old_state| {
+                    Box::pin(Self::transition_funded(
+                        old_state,
+                        keypair,
+                        contract,
+                        dbtx,
+                        global_context.clone(),
+                    ))
+                },
+            ),
+            StateTransition::new(
+                Self::await_payment_timeout(timeout),
+                |_dbtx, (), old_state| Box::pin(Self::transition_timeout(old_state)),
+            ),
+        ]
+    }
+
+    async fn await_incoming_contract_account(
+        invoice: Invoice,
+        global_context: DynGlobalClientContext,
+    ) -> IncomingContractAccount {
+        // TODO: Get rid of polling
+        loop {
+            let contract_id = (*invoice.payment_hash()).into();
+            let contract =
+                LightningClientContext::get_incoming_contract(contract_id, global_context.clone())
+                    .await;
+
+            if let Ok(contract) = contract {
+                return contract;
+            }
+
+            sleep(Duration::from_secs(1)).await;
+        }
+    }
+
+    async fn transition_funded(
+        old_state: LightningReceiveStateMachine,
+        keypair: KeyPair,
+        contract: IncomingContractAccount,
+        dbtx: &mut ClientSMDatabaseTransaction<'_, '_>,
+        global_context: DynGlobalClientContext,
+    ) -> LightningReceiveStateMachine {
+        let outpoint = Self::claim_incoming_contract(dbtx, contract, keypair, global_context).await;
+        LightningReceiveStateMachine {
+            common: old_state.common,
+            state: LightningReceiveStates::Funded(LightningReceiveFunded { outpoint }),
+        }
+    }
+
+    async fn claim_incoming_contract(
+        dbtx: &mut ClientSMDatabaseTransaction<'_, '_>,
+        contract: IncomingContractAccount,
+        keypair: KeyPair,
+        global_context: DynGlobalClientContext,
+    ) -> OutPoint {
+        let input = contract.claim();
+        let client_input = ClientInput::<LightningInput, LightningClientStateMachines> {
+            input,
+            keys: vec![keypair],
+            // The input of the refund tx is managed by this state machine, so no new state machines
+            // need to be created
+            state_machines: Arc::new(|_, _| vec![]),
+        };
+
+        let txid = global_context.claim_input(dbtx, client_input).await;
+        OutPoint { txid, out_idx: 0 }
+    }
+
+    async fn await_payment_timeout(timeout: Duration) {
+        sleep(timeout).await
```

Comment:
We'll probably want to add some buffer time since clocks tend to not be perfectly synchronized and we'd rather wait for payment for too long, making some useless API requests, than lose money.

Reply:
Added 10% buffer to the existing timeout